src/parse.rs

```diff
@@ -0,0 +1,2256 @@
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! [Cucumber Expressions][1] [AST][2] parser.
+//!
+//! See details in the [grammar spec][3].
+//!
+//! [1]: https://github.com/cucumber/cucumber-expressions#readme
+//! [2]: https://en.wikipedia.org/wiki/Abstract_syntax_tree
+//! [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+
+use std::{fmt::Display, ops::RangeFrom};
+
+use derive_more::{Display, Error};
+use nom::{
+    branch::alt,
+    bytes::complete::{tag, take_while},
+    character::complete::one_of,
+    combinator::{map, peek, verify},
+    error::{ErrorKind, ParseError},
+    multi::{many0, many1, separated_list1},
+    sequence::tuple,
+    AsChar, Compare, Err, FindToken, IResult, InputIter, InputLength,
+    InputTake, InputTakeAtPosition, Needed, Offset, Parser, Slice,
+};
+
+use crate::{
+    ast::{
+        Alternation, Alternative, Expression, Optional, Parameter,
+        SingleExpression,
+    },
+    combinator,
+};
+
+/// Reserved characters requiring special handling.
+pub const RESERVED_CHARS: &str = r#"{}()\/ "#;
+
+/// Matches `normal` and [`RESERVED_CHARS`] escaped with `\`.
+///
+/// Uses [`combinator::escaped0`] under the hood.
+///
+/// # Errors
+///
+/// ## Recoverable [`Error`]
+///
+/// - If `normal` parser errors.
+///
+/// ## Irrecoverable [`Failure`]
+///
+/// - If `normal` parser fails
+/// - [`EscapedNonReservedCharacter`]
+///
+/// [`Error`]: Err::Error
+/// [`EscapedNonReservedCharacter`]: Error::EscapedNonReservedCharacter
+/// [`Failure`]: Err::Failure
+fn escaped_reserved_chars0<'a, Input: 'a, F, O1>(
+    normal: F,
+) -> impl FnMut(Input) -> IResult<Input, Input, Error<Input>>
+where
+    Input: Clone
+        + Display
+        + Offset
+        + InputLength
+        + InputTake
+        + InputTakeAtPosition
+        + Slice<RangeFrom<usize>>
+        + InputIter,
+    <Input as InputIter>::Item: AsChar + Copy,
+    F: Parser<Input, O1, Error<Input>>,
+    Error<Input>: ParseError<Input>,
+    for<'s> &'s str: FindToken<<Input as InputIter>::Item>,
+{
+    combinator::map_err(
+        combinator::escaped0(normal, '\\', one_of(RESERVED_CHARS)),
+        |e| {
+            if let Err::Error(Error::Other(span, ErrorKind::Escaped)) = e {
+                let span = (span.input_len() > 0)
+                    .then(|| span.take(1))
+                    .unwrap_or(span);
+                Error::EscapedNonReservedCharacter(span).failure()
+            } else {
+                e
+            }
+        },
+    )
+}
+
+/// # Syntax
+///
+/// ```text
+/// parameter       := '{' (name | '\' name_to_escape)* '}'
+/// name            := ^name_to_escape
+/// name_to_escape  := '{' | '}' | '(' | '/' | '\'
```

Comment:
It's better to preserve here the original EBNF grammar syntax, so it will match the architecture document, and preserve original naming asap.

Markdown can do EBNF pretty well, though:
```ebnf
parameter = "{" (name | "\" name_to_escape)* "}"
```

Reply:
As described above, provided EBNF grammar tries to describe language that is wider than `Cucumber Expression`, but has errors. Grammar in my PR describes exactly `Cucumber Expression`.